frame-metadata/src/lib.rs

```diff
@@ -19,349 +19,28 @@
-#[derive(Clone, PartialEq, Eq, Encode)]
-#[cfg_attr(feature = "std", derive(Decode, Serialize, Debug))]
-pub struct EventMetadata {
-	pub name: DecodeDifferentStr,
-	pub arguments: DecodeDifferentArray<&'static str, StringBuf>,
-	pub documentation: DecodeDifferentArray<&'static str, StringBuf>,
-}
-
-/// All the metadata about one storage entry.
-#[derive(Clone, PartialEq, Eq, Encode)]
-#[cfg_attr(feature = "std", derive(Decode, Serialize, Debug))]
-pub struct StorageEntryMetadata {
-	pub name: DecodeDifferentStr,
-	pub modifier: StorageEntryModifier,
-	pub ty: StorageEntryType,
-	pub default: ByteGetter,
-	pub documentation: DecodeDifferentArray<&'static str, StringBuf>,
-}
-
-/// All the metadata about one module constant.
-#[derive(Clone, PartialEq, Eq, Encode)]
-#[cfg_attr(feature = "std", derive(Decode, Serialize, Debug))]
-pub struct ModuleConstantMetadata {
-	pub name: DecodeDifferentStr,
-	pub ty: DecodeDifferentStr,
-	pub value: ByteGetter,
-	pub documentation: DecodeDifferentArray<&'static str, StringBuf>,
-}
-
-/// All the metadata about a module error.
-#[derive(Clone, PartialEq, Eq, Encode)]
-#[cfg_attr(feature = "std", derive(Decode, Serialize, Debug))]
-pub struct ErrorMetadata {
-	pub name: DecodeDifferentStr,
-	pub documentation: DecodeDifferentArray<&'static str, StringBuf>,
-}
-
-/// All the metadata about errors in a module.
-pub trait ModuleErrorMetadata {
-	fn metadata() -> &'static [ErrorMetadata];
-}
-
-impl ModuleErrorMetadata for &'static str {
-	fn metadata() -> &'static [ErrorMetadata] {
-		&[]
-	}
-}
-
-/// A technical trait to store lazy initiated vec value as static dyn pointer.
-pub trait DefaultByte: Send + Sync {
-	fn default_byte(&self) -> Vec<u8>;
-}
-
-/// Wrapper over dyn pointer for accessing a cached once byte value.
-#[derive(Clone)]
-pub struct DefaultByteGetter(pub &'static dyn DefaultByte);
-
-/// Decode different for static lazy initiated byte value.
-pub type ByteGetter = DecodeDifferent<DefaultByteGetter, Vec<u8>>;
-
-impl Encode for DefaultByteGetter {
-	fn encode_to<W: Output>(&self, dest: &mut W) {
-		self.0.default_byte().encode_to(dest)
-	}
-}
-
-impl codec::EncodeLike for DefaultByteGetter {}
-
-impl PartialEq<DefaultByteGetter> for DefaultByteGetter {
-	fn eq(&self, other: &DefaultByteGetter) -> bool {
-		let left = self.0.default_byte();
-		let right = other.0.default_byte();
-		left.eq(&right)
-	}
-}
-
-impl Eq for DefaultByteGetter {}
-
-#[cfg(feature = "std")]
-impl serde::Serialize for DefaultByteGetter {
-	fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
-	where
-		S: serde::Serializer,
-	{
-		self.0.default_byte().serialize(serializer)
-	}
-}
+use codec::{Encode, Output};
 
-impl core::fmt::Debug for DefaultByteGetter {
-	fn fmt(&self, f: &mut core::fmt::Formatter) -> core::fmt::Result {
-		self.0.default_byte().fmt(f)
+cfg_if::cfg_if! {
+	if #[cfg(feature = "v12")] {
+		mod v12;
+		pub use v12::*;
+	} else if #[cfg(feature = "v13")] {
```

Comment:
```suggestion
	}
	if #[cfg(feature = "v13")] {
```
It should be possible to enable both `v12` and `v13

Reply:
Right, we will need to reconsider the exports e.g. `pub use v12::*;` because there will be some duplicates. 

Can remove those and just have the user import the `v12` `v13` types as required?

Reply:
Yes, let's avoid the reexports and be explicit about versions.

Reply:
Done, please have a look and merge if happy